File: packages/core/src/handheld.ts
import { magnitude, severityFromPeak } from "./geo.js";
import { harshEventLevel } from "./harsh.js";
import { gravityTiltDeg, sampleLinearAccel } from "./impact.js";
import type { DrivingEvent, ImuSample, LocationSample, Vec3 } from "./types.js";

export const PHONE_HANDHELD_TYPE = "phone_handheld" as const;

/** Gravity tilt from the mount baseline that starts a pickup (degrees). */
export const DEFAULT_HANDHELD_TILT_DEG = 35;
/** Close the span when tilt stays below this. */
export const DEFAULT_HANDHELD_EXIT_TILT_DEG = 18;
/** Gyro magnitude that counts as handling (~70°/s). */
export const DEFAULT_HANDHELD_GYRO_RADPS = 1.2;
/** Linear-accel magnitude that counts as handling. */
export const DEFAULT_HANDHELD_MOTION_MPS2 = 2;
/** Quiet enough to refresh the mount baseline. */
export const DEFAULT_HANDHELD_QUIET_GYRO_RADPS = 0.25;
export const DEFAULT_HANDHELD_QUIET_MOTION_MPS2 = 0.8;
/** How long the phone must stay quiet before the baseline locks / updates. */
export const DEFAULT_HANDHELD_STABLE_MS = 800;
/** Pickup must stay tilted + handling this long before opening a span. */
export const DEFAULT_HANDHELD_CONFIRM_MS = 350;
/** How long tilt must stay below exit (or the car stopped) before closing. */
export const DEFAULT_HANDHELD_EXIT_MS = 700;
/** Gap after a closed span before another can open. */
export const DEFAULT_HANDHELD_COOLDOWN_MS = 2500;
/** EMA toward gravity while the phone is docked and quiet. */
export const DEFAULT_HANDHELD_BASELINE_ALPHA = 0.08;

export type HandheldDetectorConfig = {
  handheldTiltDeg: number;
  handheldExitTiltDeg: number;
  handheldGyroRadps: number;
  handheldMotionMps2: number;
  handheldQuietGyroRadps: number;
  handheldQuietMotionMps2: number;
  handheldStableMs: number;
  handheldConfirmMs: number;
  handheldExitMs: number;
  handheldCooldownMs: number;
  handheldBaselineAlpha: number;
  jerkSettleMs: number;
  harshMediumX: number;
  harshHeavyX: number;
};

export type HandheldFilter = {
  baseline: Vec3 | null;
  quietSince: number | null;
  candidateSince: number | null;
  belowExitSince: number | null;
};

export function emptyHandheldFilter(): HandheldFilter {
  return {
    baseline: null,
    quietSince: null,
    candidateSince: null,
    belowExitSince: null,
  };
}

function copyVec(vector: Vec3): Vec3 {
  return { x: vector.x, y: vector.y, z: vector.z };
}

function blendBaseline(previous: Vec3 | null, gravity: Vec3, alpha: number): Vec3 {
  if (!previous) {
    return copyVec(gravity);
  }
  return {
    x: previous.x + (gravity.x - previous.x) * alpha,
    y: previous.y + (gravity.y - previous.y) * alpha,
    z: previous.z + (gravity.z - previous.z) * alpha,
  };
}

export function isPhoneHandheld(
  event: Pick<DrivingEvent, "type">,
): event is DrivingEvent & { type: "phone_handheld" } {
  return event.type === PHONE_HANDHELD_TYPE;
}

export type HandheldStep = {
  filter: HandheldFilter;
  open: DrivingEvent | null;
  /** Newly opened, updated, or closed span (hosts stream this). */
  emitted: DrivingEvent | null;
  lastClosedAt: number | undefined;
};

/**
 * Mount baseline + pickup/hold while the vehicle is moving.
 * Open on sustained gravity tilt with handling motion; stay open while tilted;
 * close when the phone returns toward the mount (or the car stops).
 * Peak is gravity tilt in degrees. Not a score event.
 */
export function advanceHandheld(args: {
  filter: HandheldFilter;
  open: DrivingEvent | null;
  lastClosedAt: number | undefined;
  sample: ImuSample;
  moving: boolean;
  speedMps: number | null;
  location: LocationSample | null;
  startedAtMs: number;
  config: HandheldDetectorConfig;
}): HandheldStep {
  const { sample, config } = args;
  const gravity = sample.gravity;
  const filter: HandheldFilter = { ...args.filter };
  let open = args.open;
  let lastClosedAt = args.lastClosedAt;
  let emitted: DrivingEvent | null = null;

  if (!gravity || magnitude(gravity) < 0.5) {
    return { filter, open, emitted, lastClosedAt };
  }

  const gyroMag = sample.gyro ? magnitude(sample.gyro) : 0;
  const linearMag = magnitude(sampleLinearAccel(sample));
  const quiet =
    gyroMag < config.handheldQuietGyroRadps && linearMag < config.handheldQuietMotionMps2;

  if (!open && quiet) {
    if (filter.quietSince == null) {
      filter.quietSince = sample.t;
    }
    if (sample.t - filter.quietSince >= config.handheldStableMs) {
      filter.baseline = blendBaseline(
        filter.baseline,
        gravity,
        filter.baseline ? config.handheldBaselineAlpha : 1,
      );
    }
  } else if (!open) {
    filter.quietSince = null;
  }

  const settled = sample.t - args.startedAtMs >= config.jerkSettleMs;
  if (!settled || !filter.baseline) {
    filter.candidateSince = null;
    return { filter, open, emitted, lastClosedAt };
  }

  const tilt = gravityTiltDeg(filter.baseline, gravity);

  if (open) {
    if (tilt > open.peak) {
      open.peak = tilt;
      open.severity = severityFromPeak(tilt, config.handheldTiltDeg);
      open.level = harshEventLevel(
        tilt,
        config.handheldTiltDeg,
        config.harshMediumX,
        config.harshHeavyX,
      );
      open.lat = args.location?.lat ?? open.lat;
      open.lon = args.location?.lon ?? open.lon;
      open.speedMps = args.speedMps;
      emitted = open;
    }
    const shouldExit = !args.moving || tilt < config.handheldExitTiltDeg;
    if (shouldExit) {
      if (filter.belowExitSince == null) {
        filter.belowExitSince = sample.t;
      }
      if (sample.t - filter.belowExitSince >= config.handheldExitMs) {
        open.endT = sample.t;
        lastClosedAt = sample.t;
        emitted = open;
        open = null;
        filter.candidateSince = null;
        filter.belowExitSince = null;
        filter.quietSince = null;
      }
    } else {
      filter.belowExitSince = null;
    }
    return { filter, open, emitted, lastClosedAt };
  }

  if (!args.moving) {
    filter.candidateSince = null;
    filter.belowExitSince = null;
    return { filter, open, emitted, lastClosedAt };
  }

  if (
    lastClosedAt != null &&
    sample.t - lastClosedAt < config.handheldCooldownMs
  ) {
    filter.candidateSince = null;
    return { filter, open, emitted, lastClosedAt };
  }

  const handling =
    gyroMag >= config.handheldGyroRadps || linearMag >= config.handheldMotionMps2;
  if (tilt >= config.handheldTiltDeg && handling && filter.candidateSince == null) {
    filter.candidateSince = sample.t;
  }
  if (filter.candidateSince != null && tilt >= config.handheldTiltDeg) {
    if (sample.t - filter.candidateSince >= config.handheldConfirmMs) {
      open = {
        id: `${PHONE_HANDHELD_TYPE}-${sample.t}`,
        type: PHONE_HANDHELD_TYPE,
        t: filter.candidateSince,
        endT: null,
        peak: tilt,
        severity: severityFromPeak(tilt, config.handheldTiltDeg),
        level: harshEventLevel(
          tilt,
          config.handheldTiltDeg,
          config.harshMediumX,
          config.harshHeavyX,
        ),
        lat: args.location?.lat ?? null,
        lon: args.location?.lon ?? null,
        speedMps: args.speedMps,
        overlaps: [],
      };
      emitted = open;
      filter.candidateSince = null;
      filter.belowExitSince = null;
    }
  } else {
    filter.candidateSince = null;
  }

  return { filter, open, emitted, lastClosedAt };
}

export function closeHandheldSpan(
  open: DrivingEvent | null,
  t: number,
): DrivingEvent | null {
  if (!open || open.endT != null) {
    return null;
  }
  open.endT = t;
  return open;
}

File: packages/native/android/src/main/java/com/harshy/sdk/Handheld.kt
package com.harshy.sdk

internal const val DEFAULT_HANDHELD_TILT_DEG = 35.0
internal const val DEFAULT_HANDHELD_EXIT_TILT_DEG = 18.0
internal const val DEFAULT_HANDHELD_GYRO_RADPS = 1.2
internal const val DEFAULT_HANDHELD_MOTION_MPS2 = 2.0
internal const val DEFAULT_HANDHELD_QUIET_GYRO_RADPS = 0.25
internal const val DEFAULT_HANDHELD_QUIET_MOTION_MPS2 = 0.8
internal const val DEFAULT_HANDHELD_STABLE_MS = 800.0
internal const val DEFAULT_HANDHELD_CONFIRM_MS = 350.0
internal const val DEFAULT_HANDHELD_EXIT_MS = 700.0
internal const val DEFAULT_HANDHELD_COOLDOWN_MS = 2500.0
internal const val DEFAULT_HANDHELD_BASELINE_ALPHA = 0.08

const val PHONE_HANDHELD_TYPE = "phone_handheld"

fun isPhoneHandheld(event: DrivingEvent): Boolean = event.type == PHONE_HANDHELD_TYPE

internal data class HandheldFilter(
  var baseline: Vec3? = null,
  var quietSince: Double? = null,
  var candidateSince: Double? = null,
  var belowExitSince: Double? = null,
)

internal data class HandheldStep(
  val filter: HandheldFilter,
  val open: DrivingEvent?,
  val emitted: DrivingEvent?,
  val lastClosedAt: Double?,
)

private fun blendBaseline(previous: Vec3?, gravity: Vec3, alpha: Double): Vec3 {
  if (previous == null) {
    return copyVec(gravity)
  }
  return Vec3(
    x = previous.x + (gravity.x - previous.x) * alpha,
    y = previous.y + (gravity.y - previous.y) * alpha,
    z = previous.z + (gravity.z - previous.z) * alpha,
  )
}

internal fun emptyHandheldFilter(): HandheldFilter = HandheldFilter()

internal fun advanceHandheld(
  filter: HandheldFilter,
  open: DrivingEvent?,
  lastClosedAt: Double?,
  sample: ImuSample,
  moving: Boolean,
  speedMps: Double?,
  location: LocationSample?,
  startedAtMs: Double,
  config: DetectorConfig,
): HandheldStep {
  val gravity = sample.gravity
  var nextOpen = open
  var nextClosedAt = lastClosedAt
  var emitted: DrivingEvent? = null
  val next = filter.copy(
    baseline = filter.baseline?.let { copyVec(it) },
  )

  if (gravity == null || magnitude(gravity) < 0.5) {
    return HandheldStep(next, nextOpen, emitted, nextClosedAt)
  }

  val gyroMag = sample.gyro?.let { magnitude(it) } ?: 0.0
  val linearMag = magnitude(sampleLinearAccel(sample))
  val quiet =
    gyroMag < config.handheldQuietGyroRadps && linearMag < config.handheldQuietMotionMps2

  if (nextOpen == null && quiet) {
    if (next.quietSince == null) {
      next.quietSince = sample.t
    }
    if (sample.t - (next.quietSince ?: sample.t) >= config.handheldStableMs) {
      next.baseline = blendBaseline(
        next.baseline,
        gravity,
        if (next.baseline == null) 1.0 else config.handheldBaselineAlpha,
      )
    }
  } else if (nextOpen == null) {
    next.quietSince = null
  }

  val settled = sample.t - startedAtMs >= config.jerkSettleMs
  if (!settled || next.baseline == null) {
    next.candidateSince = null
    return HandheldStep(next, nextOpen, emitted, nextClosedAt)
  }

  val tilt = gravityTiltDeg(next.baseline, gravity)

  if (nextOpen != null) {
    val openEvent = nextOpen
    if (tilt > openEvent.peak) {
      openEvent.peak = tilt
      openEvent.severity = severityFromPeak(tilt, config.handheldTiltDeg)
      openEvent.level = harshEventLevel(
        tilt,
        config.handheldTiltDeg,
        config.harshMediumX,
        config.harshHeavyX,
      )
      openEvent.lat = location?.lat ?: openEvent.lat
      openEvent.lon = location?.lon ?: openEvent.lon
      openEvent.speedMps = speedMps
      emitted = openEvent
    }
    val shouldExit = !moving || tilt < config.handheldExitTiltDeg
    if (shouldExit) {
      if (next.belowExitSince == null) {
        next.belowExitSince = sample.t
      }
      if (sample.t - (next.belowExitSince ?: sample.t) >= config.handheldExitMs) {
        openEvent.endT = sample.t
        nextClosedAt = sample.t
        emitted = openEvent
        nextOpen = null
        next.candidateSince = null
        next.belowExitSince = null
        next.quietSince = null
      }
    } else {
      next.belowExitSince = null
    }
    return HandheldStep(next, nextOpen, emitted, nextClosedAt)
  }

  if (!moving) {
    next.candidateSince = null
    next.belowExitSince = null
    return HandheldStep(next, nextOpen, emitted, nextClosedAt)
  }

  if (nextClosedAt != null && sample.t - nextClosedAt < config.handheldCooldownMs) {
    next.candidateSince = null
    return HandheldStep(next, nextOpen, emitted, nextClosedAt)
  }

  val handling =
    gyroMag >= config.handheldGyroRadps || linearMag >= config.handheldMotionMps2
  if (tilt >= config.handheldTiltDeg && handling && next.candidateSince == null) {
    next.candidateSince = sample.t
  }
  if (next.candidateSince != null && tilt >= config.handheldTiltDeg) {
    if (sample.t - (next.candidateSince ?: sample.t) >= config.handheldConfirmMs) {
      val opened = DrivingEvent(
        id = "$PHONE_HANDHELD_TYPE-${sample.t}",
        type = PHONE_HANDHELD_TYPE,
        t = next.candidateSince ?: sample.t,
        endT = null,
        peak = tilt,
        severity = severityFromPeak(tilt, config.handheldTiltDeg),
        level = harshEventLevel(
          tilt,
          config.handheldTiltDeg,
          config.harshMediumX,
          config.harshHeavyX,
        ),
        lat = location?.lat,
        lon = location?.lon,
        speedMps = speedMps,
      )
      nextOpen = opened
      emitted = opened
      next.candidateSince = null
      next.belowExitSince = null
    }
  } else {
    next.candidateSince = null
  }

  return HandheldStep(next, nextOpen, emitted, nextClosedAt)
}

internal fun closeHandheldSpan(open: DrivingEvent?, t: Double): DrivingEvent? {
  if (open == null || open.endT != null) {
    return null
  }
  open.endT = t
  return open
}
